drill.cpp

```diff
@@ -549,13 +557,48 @@ bool ExcellonProcessor::millhole(std::ofstream &of, double start_x, double start
           }
         }
 
-        of << "G1 Z" << cutter->zsafe * cfactor
+        of << "G1 Z" << return_z * cfactor
            << " F" << cutter->vertfeed * cfactor << "\n\n";
 
         return true;
     }
 }
 
+// mill a hole im multiple passes; returns false if tool is bigger than the circle
+bool ExcellonProcessor::millhole(std::ofstream &of, double start_x, double start_y,
+                                 double stop_x, double stop_y,
+                                 shared_ptr<Cutter> cutter,
+                                 double holediameter) {
+  // mill a holediameter size hole in multiple steps, starting from a size
+  // between min_entry and max_entry, enlarging by less than diameter_step each
+  // time
+  double min_entry = cutter->tool_diameter * min_milldrill_entry_diameter.asFraction(1);
+  double max_entry = cutter->tool_diameter * max_milldrill_entry_diameter.asFraction(1);
+  double diameter_step = cutter->tool_diameter * milldrill_stepover.asFraction(1) * 2.0;
+
+  if (holediameter < max_entry) {
+    // can do this in one pass
+    return millhole_one(of, start_x, start_y, stop_x, stop_y, cutter, holediameter);
+  } else {
+    // number of diameter_step size enlargements to go from a size under
+    // max_entry to holediameter without exceeding diameter_step
+    int steps = (int)std::ceil((holediameter - max_entry) / diameter_step);
+
+    // adjust entry diameter and reduce step size to meet min_entry requirements
+    double entry_diameter = std::max(
+        holediameter - steps * diameter_step,
+        min_entry);
+    diameter_step = (holediameter - entry_diameter) / steps;
+
+    for (int step = 0; step <= steps; step++) {
+      millhole_one(of, start_x, start_y, stop_x, stop_y, cutter,
+                   entry_diameter + step * diameter_step, step == steps);
+    }
+
+    return true;
```

Comment:
The only thing this is used for is to print this warning https://github.com/pcb2gcode/pcb2gcode/blob/b1dade8f669e6441dca613539a4ab7b606831e66/drill.cpp#L716-L717

If the rest of the logic is correct (and the parameters are validated), returning true is correct because that bit of code always cuts multiple passes, and there is no reason to enlarge a hole that has already been cut too big.

I also wasn't sure what to do with this condition. As it is I think it makes more sense for `milldrill_one` to return `void`, and move the condition to `milldrill`. Does that make sense?

An alternative would be just returning the last status returned by `milldrill`.

It could be higher still, but i can see some logic to deciding this in the `milldrill` function, for example if you decide to do something smart, like enlarging circular holes (so that they are never plunged), but allowing slot milling (which makes its own clearance), that would be the right place for it. 

Alternatively, the logic for why it's "correct" is:
- in `milldrill_one`, the condition for returning false is `cutter->tool_diameter * 1.001 >= holediameter`
- so the true condition is  `cutter->tool_diameter * 1.001 < holediameter`
- in milldrill, the true condition is: `max_entry <= holediameter `
- which is equal to `cutter->tool_diameter * max_milldrill_entry_diameter <= holediameter `
- `max_milldrill_entry_diameter` should be >= 100% (though in reality it would be odd to set it that low), so the conditions are nearly the same.

... but the existing condition is not exactly right, because it will return false (and issue a warning) if using a 1mm cutter to mill a 1mm slot, which is perfectly valid. I think we should fix this while we're thinking about it and understand it.

Reply:
Right, that tolerance is required to not produce very small arcs, but the condition for the warning should a bit different, i think.

Currently it works like this for a 1mm cutter:

- holesize >= 1.001mm: use arcs, no warning
- holesize < 1.001mm: no arc, issue warning

The warning is not quite right if holesize = 1.0mm

Instead it should be:

- holesize >= 1.001mm: use arcs
- holesize < 1.001mm: no arcs
- holesize >= 0.999mm: no warning
- holesize < 0.999mm: no warning

This way, a 1mm hole doesn't use arcs, but also doesn't get a warning.